rotkehlchen/history/events/structures/swap.py

```diff
@@ -0,0 +1,222 @@
+        )
+
+    def serialize(self) -> dict[str, Any]:
+        """Serialize the event for api.
+        Autogenerates the event notes, appending any notes added by the user in a second sentence.
+        """
+        serialized_data = super().serialize()
+        location_name = get_formatted_location_name(self.location)
+        asset_symbol = self.asset.resolve_to_asset_with_symbol().symbol
+        if self.event_subtype == HistoryEventSubType.SPEND:
+            notes = f'Swap {self.amount} {asset_symbol} in {location_name}.'
+        elif self.event_subtype == HistoryEventSubType.RECEIVE:
+            notes = f'Receive {self.amount} {asset_symbol} after a swap in {location_name}.'
+        else:  # Fee
+            notes = f'Spend {self.amount} {asset_symbol} as {location_name} swap fee.'
+
+        if (user_notes := serialized_data['notes']) is not None:
+            notes += f' {user_notes}'
+
+        serialized_data['notes'] = notes
+        return serialized_data
+
+    @classmethod
+    def deserialize(cls: type['SwapEvent'], data: dict[str, Any]) -> 'SwapEvent':
+        base_data = cls._deserialize_base_history_data(data)
+        if (event_subtype := base_data['event_subtype']) not in {
+            HistoryEventSubType.SPEND,
+            HistoryEventSubType.RECEIVE,
+            HistoryEventSubType.FEE,
+        }:
+            raise DeserializationError(
+                f'Unsupported swap event subtype {event_subtype}. '
+                f'Expected SPEND, RECEIVE or FEE',
+            )
+
+        return cls(
+            identifier=base_data['identifier'],
+            event_identifier=base_data['event_identifier'],
+            timestamp=base_data['timestamp'],
+            location=base_data['location'],
+            location_label=base_data['location_label'],
+            event_subtype=event_subtype,  # type: ignore  # just confirmed it's a SPEND, RECEIVE or FEE above
+            asset=base_data['asset'],
+            amount=base_data['amount'],
+        )
+
+    def __repr__(self) -> str:
+        return f'SwapEvent({", ".join(self._history_base_entry_repr_fields())})'
+
+    # -- Methods of AccountingEventMixin
+
+    @staticmethod
+    def get_accounting_event_type() -> AccountingEventType:
+        return AccountingEventType.TRADE
+
+    def process(
+            self,
+            accounting: 'AccountingPot',
+            events_iterator: "peekable['AccountingEventMixin']",  # pylint: disable=unused-argument
+    ) -> int:
+        return accounting.events_accountant.process(self, events_iterator)
+
+
+def create_swap_events(
+        timestamp: TimestampMS,
+        location: Location,
+        spend_asset: Asset,
+        spend_amount: 'FVal',
+        receive_asset: Asset,
+        receive_amount: 'FVal',
+        fee_asset: Asset,
+        fee_amount: 'FVal',
+        location_label: str | None = None,
+        unique_id: str | None = None,
+        spend_notes: str | None = None,
```

Comment:
as I mentioned in the other part of the code perhaps all notes can be a list. Same as for asset movements